doc: clarify module system selection (#41383)
nodejs/node: doc/api/packages.md

## Determining module system

Node.js will treat the following as [ES modules][] when passed to `node` as the
initial input, or when referenced by `import` statements or `import()`
expressions:

> * Files with a `.mjs` extension.

Reviewer: ```suggestion
* Files with an `.mjs` extension.
```

Author: Hum I read that as `a dot M-J-S extension`, is there another name for `.`?

Reviewer: The periods aren’t spoken in a sentence like this. In spoken English you’d say “an em jay ess extension.” This also tracks that I would think that most people consider the extension to be `mjs`, not `.mjs` (as in, the dot isn’t part of the extension) which might explain why the period isn’t spoken.

preserve backward compatibility. However, now that Node.js supports both
CommonJS and ES modules, it is best to be explicit whenever possible. Node.js
will treat the following as CommonJS when passed to `node` as the initial input,
or when referenced by `import` statements, `import()` expressions, or

> `require()` expressions:

Reviewer: What happens when you `require` a `.js` file in a `"type": "module"` scope? Does it error the way it does for `.mjs` files, or does it try to load the `.js` file as CommonJS in spite of the `"type"` field?

Author: It throws an `ERR_REQUIRE_ESM` error, same as with `.mjs` files.

```sh
mkdir repro
cd repro
touch file.js
node -e 'require("./file.js")'
echo '{"type":"module"}' > package.json
node -e 'require("./file.js")' # ERR_REQUIRE_ESM
```

Reviewer: Excellent, I was hoping that’s what happened. 😌

it will also make things easier for build tools and loaders to determine how the
files in the package should be interpreted.

> Node.js will refuse to load the following when passed to `node` as the
> initial input and the nearest parent `package.json` file contains a top-level
> [`"type"`][] field with a value of `"module"`:
> 
> * Files whose name doesn't end in `.js`, `.mjs`, `.cjs`, or `.wasm`.

Reviewer: Rather than saying what Node _won’t_ do, why not say what it will? I think it’s easier to understand the positive case than the negative:

```suggestion
When the nearest parent `package.json` file contains a top-level
[`"type"`][] field with a value of `"module"`, the `node` command will
accept as input only files with `.js`, `.mjs`, or `.cjs` extensions;
and with `.wasm` extensions when `--experimental-wasm-modules` is enabled.
```

(And please add a link to `--experimental-wasm-modules`; and is a wasm entry point actually supported?)

Reviewer 2: The type field has no bearing on what node accepts - it accepts a .js file regardless, and has no effect on the others.

Reviewer 3: That isn't true, type affects if the ESM module loader is used for the entry point: https://github.com/nodejs/node/blob/2cc7a91a5d855b4ff78f21f1bb8d4e55131d0615/lib/internal/modules/run_main.js#L41 . Also it makes extension-less files fail to load and that should probably be called out.

Reviewer 2: For the latter, that’s good to know, and should indeed be called out.

For the former, are you saying `node foo` won’t do any extension lookup on `foo.cjs`,`foo.js`, or `foo.mjs`? If so, i understand why - but that’s not enabling ESM, that’s disabling the standard/default CJS loader, and that should definitely be explicitly called out as well (as a potential hazard/impact of using type module)

Reviewer 3: under `type:"module"` the regular ESM semantics are applied against the preresolved filepath `argv[1]` absolute `file:` URL; so, no searching occurs. For `type:"commonjs"` the CJS semantics are applied against the preresolved `argv[1]` absolute file path.

Reviewer 2: Thanks, TIL a second effect of type module :-) please do call this out explicitly; the more users correctly understand type module’s effects (which as i now understand it are just that .jjs is ESM + node CLI stops doing extension/index searching) the better.

Author: > under `type:"module"` the regular ESM semantics are applied against the preresolved filepath `argv[1]` absolute `file:` URL; so, no searching occurs.

I beg to differ:

```sh
mkdir repro
cd repro
echo 'console.log(this === undefined ? "ESM" : "CJS")' > file.js
node file # CJS
echo '{"type":"module"}' > package.json
node file # ESM
```

Reviewer 3: ooo, this one is fun. it apparently got carved out some weird edge case that I missed but it doesn't do *all* kinds of searching? https://github.com/nodejs/node/blob/master/lib/internal/modules/esm/resolve.js#L1000 doesn't seem to be the path it is taking but it is hitting shouldUseESMLoader. Is this a bug?

Reviewer 2: Seems more like a feature than a bug :-) removing it will likely be a breaking change, as users of node 12, 14, 16, and 17 are likely relying on it.

Reviewer 3: so it looks like CJS resolver is always hit prior to the the ESM resolver, but things are weird and spaghetti https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/lib/internal/modules/run_main.js#L70 , so it does both, not one or the other when it thinks it is ESM. there is an odd comment about back compat, but I have no memory of a compat issue.

Reviewer 3: @ljharb the comments in the file that are causing it explicitly state that removing extension searching can be done in the future.

Reviewer 3: also, this seems like it would break if someone mucked with require.extensions in weird ways and passed it to ESM

Reviewer 2: The comments were probably from when it was still experimental. Now that it no longer is, it doesn't seem like something that can be removed without a semver-major.

Reviewer 3: Likely would be semver-major yes.

> 

Reviewer: This is perhaps where we should mention that ESM entry points need to be URLs.

```suggestion

When the initial input is ES module JavaScript, the reference to it must be
given as a URL, not a file path. For example: `node entry.mjs`;
`node ./app/entry.mjs`; `node file:///c:/path/to/app/entry.mjs`.
```

I feel like others can come up with better wording; my point is just that something along these lines should be part of the docs, either in the ESM or CLI docs or probably both.

Author: That's.. not true? AFAIK you can only specify a path as CLI entry point, using a URL only works if it happens to coincide with a path.

```console
$ cd /tmp
$ touch file.mjs
$ node file:///tmp/file.mjs
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module '/private/tmp/file:/file/path'
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v17.3.0
$ node ./%66ile.mjs
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module '/private/tmp/%66ile.mjs'
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v17.3.0
```

Reviewer 3: *Takes a deep breath* the CLI never accepts a filepath or a url. It is always a string specifier resolved using file semantics against the cwd *PRIOR* to doing any loading stuff https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/lib/internal/bootstrap/pre_execution.js#L105 . It would resolve the same as a URL or filepath semantically. This causes confusions since it uses file resolution, but that is something that would be too dangerous to change last I talked with others about it (percent encoding problems for example). The increased usage of URL in various places may change that, but even if the CLI was to move to be URL based it likely would need a flag, but it does use URLs when it goes into the ESM Loader.

Reviewer: So in other words, it is true?

This was what tripped me up in the tests PR, where the tests were failing only in Windows and it took me a long time to figure out why.

Author: > It would resolve the same as a URL or filepath semantically.

My understanding is it only accepts a CJS specifier, because of this `Module._findPath` call: https://github.com/nodejs/node/blob/2cc7a91a5d855b4ff78f21f1bb8d4e55131d0615/lib/internal/modules/run_main.js#L16

At least this discussion shows that this part of the docs definitely needs some clarification, because I'm quite confused right now ^^

> This was what tripped me up in the tests PR, where the tests were failing only in Windows and it took me a long time to figure out why.

I think there's a confusion with `--experimental-loader` that takes a URL and never a path (even if the loader is a `.cjs` file) – or maybe a more correct way of saying it would be: it accepts any specifier that can be resolved by ESM loader relative to cwd?

Reviewer 3: `--loader` differs greatly reading through edge case carved out code. It was always intended to be a URL for forwards compatibility and to avoid some questions about passing in an entire loader via ARGV without a file via `data:`.

Passing to `node` as the initial input when the nearest parent `package.json`
file contains a top-level [`"type"`][] field with a value of `"commonjs"`, or
when referenced by `require()` calls:

* Files whose name ends in `.node` are interpreted as
  compiled addon modules loaded with `process.dlopen()`.

* Files whose name ends in `.json` are parsed as JSON text files.

* Any other files will be treated as a [CommonJS][] module.

> 

Reviewer 4: Just wondering if we should move all this content to the CLI documentation perhaps?

it will also make things easier for build tools and loaders to determine how the
files in the package should be interpreted.

### Modules loaders

> Node.js has two system for resolving a specifier and load modules.

Reviewer: ```suggestion
Node.js has two systems for resolving a specifier and loading modules.
```

There is the CommonJS module loader:

* It is fully synchronous.
* It is monkey patchable.

> * When resolving a specifier, if no exact match is found, it will try to add
>   extensions (`.js`, `.json`, and finally `.node`).

Reviewer: ```suggestion
* When resolving a specifier, if no exact match is found, it will try to add
  extensions (`.js`, `.json`, and finally `.node`) and index files
  (`/index.js`, `/index.json`, and finally `/index.node`).
```

Author: I don't think that's true, if the specifier points to an existing directory, then an exact match is found – we have a [Folders as modules](https://nodejs.org/api/modules.html#folders-as-modules) section that explains it. Otherwise, we would have to specify it first tries to load the `"main"` field if the directory contains a `package.json` file, we might as well just link to that section.

Reviewer: Then maybe mention the folder resolution without spelling it out?

```
* When resolving a specifier, if no exact match is found, it will try to add
  extensions (`.js`, `.json`, and finally `.node`) and then attempt to resolve
  [folders as modules][].

Author: Does that work? https://github.com/nodejs/node/blob/93d6a233cf3afb504b62c6de9396e9edccd1da57/doc/api/packages.md#L97-L99

* It supports [folders as modules][].

> * It treats `.json` as JSON text files.

Reviewer: I’m not sure what this means. How else would a `.json` file be treated?

Author: As JavaScript? I mean, `.wasm` is treated as JavaScript, so I think it's worth being explicit here ^^ also it's to match the ESM list below.

Reviewer: Maybe just remove the JSON bullet point from both lists? I’m not sure we need it. The lists are focusing on the differences between the loaders, not the similarities, and they can both import JSON.

Author: ES loader currently cannot import `.json` files by default. And even if it did, I think we should still list JSON support here.

Reviewer: Okay, I can’t think of better phrasing, this is fine for now.

* `.node` files are interpreted as compiled addon modules loaded with
  `process.dlopen()`.

> * It treats files that do not have `.json` or `.node` extension as JavaScript

Reviewer: ```suggestion
* It treats all files that lack `.json` or `.node` extensions as JavaScript
```

  text files.

> * It cannot be used to load ECMAScript modules. Attempting to do so will result
>   in a [`ERR_REQUIRE_ESM`][] error.
> * It can be accessed using `require` function.

Reviewer: ```suggestion
* It allows one CommonJS module to load another using the `require` function.
* It allows CommonJS modules to asynchronously import ECMAScript modules,
  using `import()` expressions.
* `require` cannot be used to load ECMAScript modules. Attempting to do so
  will result in a [`ERR_REQUIRE_ESM`][] error.
```

Author: I don't know if you noticed, but I tried to list the same features in both lists, so user can compare the difference between CJS and ESM loaders. If we add an item here, can we add it to the ESM list as well?

Reviewer: You could combine the two `require`-related ones, maybe?

```md
* It allows one CommonJS module to load another using the `require` function.
  `require` cannot be used to load ECMAScript modules. Attempting to do so
  will result in a [`ERR_REQUIRE_ESM`][] error.
* It allows CommonJS modules to asynchronously import ECMAScript modules,
  using `import()` expressions.
```

Author: > ```lisp
> It allows one CommonJS module to load another using the `require` function.
> ```

It also allows ES module to load a CJS module using `require` function. I'm not sure this information is useful tbh. Is it correct that the information you are trying to convey is that all JS files are treated as CJS? If so sure I can add that.

> ```lisp
> It allows CommonJS modules to asynchronously import ECMAScript modules,
>   using `import()` expressions.
> ```

I don't see how this is related to the CJS loader. `import()` is exposed in CJS modules, it is explicitly called out is several places, but I think it would add too much confusion to list `import()` as a CJS loader feature.

Author: I went with https://github.com/nodejs/node/blob/21b45bf922d2768ce10d480a9c843ec20b92d9f3/doc/api/packages.md#L103-L107 wdyt?

Reviewer: The “it” here I think refers to the CommonJS _loader,_ but I think what users care about is what can be done from a CommonJS _context._ As in, you _can_ load an ES module from within a CommonJS module; you just need to use `import()`. And you can load CommonJS modules from within ES modules, either via `import` or by creating `require`. In those cross-module-system cases we sort of have one loader using the other. So while it may technically be correct that the CommonJS loader only handles CommonJS and the ESM loader only handles ESM, that implies that the two systems can’t interoperate which is _not_ an impression we want to convey.

What I was trying to get across was that `require` can only refer to CommonJS, `import` and `import()` can refer to either module system; and CommonJS supports `import()`. As in, _these are the ways that interoperability is possible._ It’s important to note that `require` doesn’t support ESM, and it’s just as important to suggest `import()` is the solution for that.

Author: Fair, I agree we don't want to add this kind of possible confusion. What do you think of this: https://github.com/nodejs/node/blob/93d6a233cf3afb504b62c6de9396e9edccd1da57/doc/api/packages.md#L105-L108

Reviewer: I guess stepping back for a second, what’s the value of stating that the CommonJS loader only loads CommonJS? That’s describing Node’s internal workings which may be technically correct, but easy to misunderstand; and maybe it’s better just left out entirely.

Maybe just include the latter part, in which case you can get closer to your original version:

```markdown
 * It tries to load all JavaScript text files as CommonJS modules. If such a file
   is an ECMAScript module, an [`ERR_REQUIRE_ESM`][] error is thrown.
   Within a CommonJS module, [`import()`][] can be used to invoke the
   ES module loader to import an ES module.
```

Author: What I don't "like" with this last version is it gives the impression that if an ESM file do not use any ESM specific syntax, CJS loader would (try to) parse it as CJS. Can we have a version without using `tries to`?

Do we really need to document here that `ERR_REQUIRE_ESM` is thrown? It's already documented in `modules.md`, I'd prefer if we leave it out to simplify that list item.

> what’s the value of stating that the CommonJS loader only loads CommonJS?

ESM loader can load more than just ESM, so it may not be obvious that CJS only loads CJS?

Reviewer: > ESM loader can load more than just ESM, so it may not be obvious that CJS only loads CJS?

Can it though? Doesn’t it just call the CommonJS loader to handle cases of `import` of CJS? Just like how the CommonJS loader calls the ESM loader to handle cases of `import()` of ESM?

We can leave out the error message, sure. But then I don’t think there’s anything left of this bullet point, and so we should just remove it. Both loaders use the other to handle the other system’s modules. The only thing worth mentioning is how `import`/`import()` can handle either module system while `require` only supports ESM. That’s a really important point that we should make if we’re not doing so already, and is much more valuable to users than a description of how the internals of the loaders work.

Reviewer 2: (`require` only supports CJS)

Author: IMHO there's a clear difference: with `Module._load` there is no way to load a file as ESM; with `esmLoader.load`, you can load a file as CJS. That's what I'm talking about, maybe it's not an information relevant for end users, but I believe it is for loader hooks authors.
`require` and `import` are user facing features, and are already documented elsewhere, it's easier if we define `require => CJS loader` and `import => ESM loader` (and it's mostly true), and I think it would be detrimental to imply that `import` and `require` behave differently depending on the parse goal of the file.

I'm going to land as is, hopefully we can improve this part of the docs in future PRs.

Reviewer: > ESM loader can load more than just ESM, so it may not be obvious that CJS only loads CJS?

But can it really? If you `import` a CommonJS file, isn’t it just passing that off to the CommonJS loader for loading? So in that sense, each loader only handles its domain.

CommonJS _modules_ load more than just CommonJS, and the inverse for ES modules, and that’s what matters to end users. Stating that the _loader_ can only handle CommonJS just adds confusion.

I think that matters to users is that `require` can only take CommonJS, while `import`/`import()` can take either. And that `import` doesn’t exist in CommonJS, but `import()` does.

All this other stuff about what the loaders can do is unnecessary detail that can confuse people. I understand the desire to keep the whole list focused on the CommonJS loader, which is why I’m thinking it’s best to just leave out this detail unless you want to make this bullet about `require` vs `import` rather than about the loader itself.

There is the ECMAScript module loader:

> * It is Asynchronous.
> * It is not monkey patchable, can be customized using [loader hooks][].
> * No extension searching, the specifier must point to the exact URL of the file.

Reviewer: ```suggestion
* It is asynchronous.
* It is not monkey patchable, but can be customized using [loader hooks][].
* Specifiers referencing files must include file extensions; Node.js will not
  search for them.
```

Reviewer 2: must include file extensions or match the "exports" field of the package being imported?

Author: I agree with Jordan, it's not correct to say specifier must include file extensions. Maybe we can reuse the wording from https://github.com/nodejs/node/blob/1b1159ae502a134c48427f6351627d32767539af/doc/api/esm.md#L149-L151

Reviewer 2: directory indexes must only be fully specified *when using relative or absolute specifiers* tho, which i'm not sure is clear enough without being part of the same sentence, or stating it twice.

Author: I went with https://github.com/nodejs/node/blob/2cea6a8fcd4ba7ccb9a7dfe632ba33b489786f42/doc/api/packages.md#L113-L114 let me know if you think it needs more work.

Reviewer: I’m fine with the terms, but that particular sentence has a comma splice, which is incorrect grammar.

```markdown
 * It does no extension searching. A file extension must be provided
   when the specifier is a relative or absolute file URL.
```

* It does not support folders as modules.

> * Import assertion are needed to load JSON modules (behind
>   `--experimental-json-modules` flag).
> * It only accepts `.js`, `.mjs`, and `.cjs` extensions for JavaScript text
>   files.
> * It can be used to load (JavaScript) CommonJS modules. It passes the module
>   content through the `es-module-lexer` to assess what are its exports, convert
>   its URL to an absolute path and load it using the CommonJS module loader.

Reviewer: ```suggestion
* It can load JSON modules, but an import assertion is required (behind
  `--experimental-json-modules` flag).
* It accepts only `.js`, `.mjs`, and `.cjs` extensions for JavaScript text
  files.
* It can be used to load JavaScript CommonJS modules. Such modules
  are passed through the `es-module-lexer` to try to identify named exports,
  which are available if they can be determined through static analysis.
  Imported CommonJS modules have their URLs converted to absolute
  paths and are then loaded via the CommonJS module loader.
```

> * It can be accessed using `import`.

Reviewer: ```suggestion
* It is responsible for handling `import` statements and `import()` expressions.
```

This should probably be near the top of the list; likewise “It is responsible for the `require` function” should be near the top of the CommonJS list.
